### ISSUE TITLE
Add Support for AWS Aurora Data API for use in AWS Lambda

### ISSUE THREAD

**Issue by WooDzu** [sequelize/sequelize#11021]
**Is your feature request related to a problem? Please describe.**
Using databases securely in AWS Lambda is cumbersome as it requires understanding of VPC networking and setting up NAT for your lambda when it needs to access both a RDS database and external resources.
It also requires managing connection pool in sequelize with has known issues and limitations.
AWS Aurora added support for Data API which is based on HTTP.
https://aws.amazon.com/about-aws/whats-new/2019/05/amazon_aurora_serverless_mysql_5_6_now_supportsdataapi/

**Describe the solution you'd like**
It would be great if sequelize could utilize AWS Aurora Data API.

**Why should this be in Sequelize**
If Sequelize directly supported it: this won't be needed any more:
- Tweaking VPC and setting up NAT etc..
- Connection pools
- packaging mysql/libmysql packages in Lambda etc..

**Describe alternatives you've considered**
* Keep using NAT/pools/libmysql
* Use Data API from SDK directly instead of sequelize

**Usage example**
* Examples for Java and Python in the documentation, nothing for JS (as of today)
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/data-api.html
* The JS SDK does seem to support it already:
https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/RDSDataService.html

----------

It might appear that the main disadvantage would be the overhead HTTP protocol,
but if you work with Lambda and RDS already the most reliable solution is disabling connection pool and initializing a fresh connection for each query - which is an overhead on its own.

**Comment by WooDzu** [comment]
It would be great if someone experienced could outline the work it will need and if a plugin/addon is more feasible than adding to the core.

**Comment by chaitanya11** [comment]
@WooDzu as per my understanding, New AWS Data Api is to manage connections on aws side (meaning, we will not have exposure to connection object). In sequelize we have to register our models after establishing successful connection, which is not possible in this case. 
please correct me if i am wrong.

**Comment by WooDzu** [comment]
I think that is true. In AWS SDK connection does not seem to be exposed and in AWS REST API (rds-data.us-east-1.amazonaws.com) there is no keep-alive or HTTP2 to utilize in connection pool.

Would that be a blocker? If this is because some form of a connection manager needs to be exposed in sequelize by design?

**Comment by chrisjhoughton** [comment]
The point of the new data API is specifically that the connection object is **not** exposed. It means that you can simply execute SQL queries via their SDKs (which are over HTTP, under the hood) without the need for a VPC.

Lambda functions have horrible cold starts when connecting to a database inside of a VPC, so the data API is something that would be hugely beneficial unless AWS rework how VPCs work in a serverless world.

In the new Data API world, essentially you just fire queries at the data API, which are then executed against the database, with the results returned to you. As such - if Sequelize were to connect to a database behind the data API, then there would be no need to even think about connections. The whole library would become a lot more "dumb".

This is such a big change to the underlying assumptions in connecting to a database, I can't help but think there would need to be a hard fork from the main Sequelize repo?

**Comment by chaitanya11** [comment]
I totally agree with @chrisjhoughton , implementation of data api needs only `Model,QueryGenerator, AWS SDK` modules, other modules from `sequelize` were not necessary to make database communication.

**Comment by KowalskiTom** [comment]
I don't think there needs to be a hard fork. I would think you just need to
override the connection methods.

**Comment by chaitanya11** [comment]
@KowalskiTom  we need to override Transactions, Sequelize initialisation (constructor), where username, host, password and some other vars are required. We need to override Query, and we need to override how to handle things like getting version (which will execute a query before executing each user query), etc... 
And limitation from aws side as well like pagination after 1MB or 100 records of data, timeout for every query operation, batch querying etc..
These were just somethings in my mind, can be more in real implementation.

**Comment by aat2703** [comment]
+1

**Comment by mattccrampton** [comment]
Checking if this is an in-progress feature for Sequelize or not.  We're moving to Aurora Serverless Data APIs and I'm trying to determine if we're going to need to move off of Sequelize as part of the work or not.

**Comment by papb** [comment]
@mattccrampton No progress that I know of, PRs always welcome

**Comment by darbio** [comment]
[TypeORM](https://typeorm.io/) supports the Data API. Maybe this could be used as a starting point for how to get this library to use it?

**Comment by cguzmancamus** [comment]
I think that the support of TypeORM is only for mysql, can anyone here confirm this?

**Comment by LoicMahieu** [comment]
@cguzmancamus TypeORM himself supports various database drivers but [`data-api-client`](https://github.com/ArsenyYankovsky/data-api-client) (which is the client used by [`typeorm-aurora-data-api-driver`](https://github.com/ArsenyYankovsky/typeorm-aurora-data-api-driver)) only supports mysql.

**Comment by luismramirezr** [comment]
Can someone explain to me an alternative way to connect a lambda to serverless Aurora (PostgreSQL) with Sequelize?

**Comment by cmendes0101** [comment]
@luismramirezr We use the RDS Proxy with Sequelize. It replaces our need for looking at the Data API alternative. Warning: RDS Proxy is currently in preview release

**Comment by shamoons** [comment]
> @luismramirezr We use the RDS Proxy with Sequelize. It replaces our need for looking at the Data API alternative. Warning: RDS Proxy is currently in preview release

Their docs say:

> You can't use RDS Proxy with Aurora Serverless clusters.
> You can't use RDS Proxy with Aurora multi-master clusters.

So I’m curious how you’re set up?

**Comment by cmendes0101** [comment]
@shamoons Not using a Serverless or multi-master cluster.

**Comment by suhasini-udayar** [comment]
anyone tried https://www.npmjs.com/package/@fyreware/mysql-data-api ? for sequelize with support for data api?

**Comment by xai1983kbu** [comment]
Hi! I just managed to create a table in a database of my aurora serverless cluster with the help of '@fyreware/mysql-data-api'.

I was getting an error - AccessDeniedException: User: arn:aws:sts::XXX... is not authorized to perform: rds-data:ExecuteStatement on resource: arn:aws:rds:us-east-1:XXX...
I added Policy 'AmazonRDSDataFullAccess' to my lambda function (full access just for a quick test, we don't need it) and it works.

I hope I will use it in the future.
Thank you @suhasini-udayar !

**Comment by ksk385** [comment]
@xai1983kbu I am trying a similar setup as you and having issues setting up my local environment.

app_1             | /project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:86
app_1             |           throw new Error(`Unable to find dialect at ${this.sequelize.config.dialectModulePath}`);
app_1             |                 ^
app_1             | Error: Unable to find dialect at node_modules/@fyreware/mysql-data-api/src/index.js
app_1             |     at ConnectionManager._loadDialectModule (/project/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:86:17)

I can see that the node_modules/@fyreware/mysql-data-api has the src/index.js file.

**Comment by xai1983kbu** [comment]
@ksk385 have you tried to use '@fyreware/mysql-data-api'? My first example with 'node_modules/@fyreware/mysql-data-api/src/index.js' is specific one because of Pulumi's way to load node's modules into lambda.

**Comment by ksk385** [comment]
@xai1983kbu That wasn't working before due to another issue but I got it working now. 

The issue for me use setting up environment variables for AWS credentials and region.

**Comment by marcogrcr** [comment]
For people experiencing issues in AWS Lambda, I have created a pull request that documents how to properly configure `sequelize` for Lambda. Hopefully you'll be able to see it in <https://sequelize.org/master/manual/aws-lambda.html> once it gets merged. In the meantime, please check #12642 out and provide feedback if you find any inaccuracies or space for improvement.

**Comment by thegnuu** [comment]
We had the same issue and started working on a package to resolve it, at least for postgres :) 
It is still WIP and the solution is only tested with postgres, but it might give you and idea how to implement it for other dialects as well.

Here is our package: https://github.com/bbitgmbh/bbit.rdsdata.postgres

Maybe this is helpful for anyone out there!

**Comment by octalpixel** [comment]
Is there any possibility to plug in  "serverless-mysql"  as a  dialectModule?

**Comment by Ranguna** [comment]
Hello @thegnuu, really nice work on that one!

I see the last commit was 3 months ago and there's still quite a bit of uncommented stuff and debug logging. How's development going?

**Comment by thegnuu** [comment]
@Ranguna it was not really my work, but thank you ;)

Yeah you are right, we are quite busy at the moment but we are using this package since 6 months in production and it works stable so far. Right now we are focusing on other stuff and cleaning up working solutions is not really our top priority. I can check it with my team and create a task to clean things up next time we have to make any changes, but I don't know how long this will take...

**Comment by mcewball13** [comment]
So, all these years later, is v7 going to help with this?

**Comment by ephys** [comment]
Adding built-in support for Data API is not currently on our backlog, but a few ongoing changes could help:

- We are separating all dialects from the core library and working on a public API to add new dialects. This means that it will be possible to create new dialects without needing to fork sequelize (such as `postgres-aurora` and `mysql-aurora`)
- Each of our officially supported dialects will continue to expose an option to replace the library we use to connect to the database. This means that the solution proposed by @thegnuu should continue to work

For built-in support, we would need to have a way to run our integration tests against that database, and to figure out how to deduplicate the code between the non-aurora and the aurora versions of the dialects.

As Data API requires a different connector library, they would need to be separate dialects instead of an option on each dialect.
